### PR TITLE
 [FIX] use std::cpp20::ostream[buf]_iterator instead of seqan3::ostream[buf]_iterator

### PR DESCRIPTION
**Pull request by marehr**
Part of #124

### CODE REVIEW

File: include/seqan3/io/stream/iterator.hpp [hunk 1]
```diff
@@ -14,19 +14,15 @@
...
-#endif // __cpp_lib_ranges
+#include <seqan3/std/algorithm>
+#include <seqan3/std/charconv>
+#include <seqan3/std/iterator>
+#include <seqan3/std/ranges>
 
 #include <seqan3/core/platform.hpp>
 #include <seqan3/core/concept/core_language.hpp>
 #include <seqan3/range/views/take.hpp>
 #include <seqan3/range/views/drop.hpp>
```
Comment (simonsasse) [inline include/seqan3/io/stream/iterator.hpp:25]:
```suggestion
#include <seqan3/range/views/drop.hpp>
#include <seqan3/range/views/take.hpp>
```

File: include/seqan3/std/iterator [hunk 2]
```diff
@@ -19,13 +19,17 @@
 #ifdef __cpp_lib_ranges
 namespace std
 {
-inline namespace cpp20
+namespace cpp20
```
Comment (rrahn) [inline include/seqan3/std/iterator:22]: removing inline was necessary, because?

Comment (marehr) [reply include/seqan3/std/iterator:22]: I normally write my reasons in the commit itself:

https://github.com/seqan/seqan3/pull/1909/commits/37dfe17eeda35476d2afb16c7824cb8e098c061f

> It turned out that inline namespace for `std::cpp20` was a bad idea
since it produced ambiguousness for `std::ostream_iterator` and
`std::cpp20::ostream_iterator`.

File: include/seqan3/io/stream/iterator.hpp [hunk 3]
```diff
@@ -381,42 +377,3 @@ class fast_ostreambuf_iterator
...
- *        constructed.
- */
-using SEQAN3_DOXYGEN_ONLY(ostream_iterator =) ::ranges::ostream_iterator;
-
-/*!\typedef seqan3::ostreambuf_iterator
- * \brief Alias for ranges::ostreambuf_iterator. Writes successive characters onto the output stream from which it
- *        was constructed.
- */
-using SEQAN3_DOXYGEN_ONLY(ostreambuf_iterator =) ::ranges::ostreambuf_iterator;
-
```
Comment (marehr) [inline include/seqan3/io/stream/iterator.hpp:419]: > Just one question to be sure. The renamed iterator types have nothing to do with our faster stream iterator implementations?

As you can see, this has nothing to do with our faster stream iterators. This is from way back when we didn't have the `seqan3/std` folder for "standard" implementations.

That means `seqan3::ostreambuf_iterator` was either `ranges::ostreambuf_iterator` (pre C++20) or `std::ostreambuf_iterator` (C++20).

File: include/seqan3/io/alignment_file/format_bam.hpp [hunk 4]
```diff
@@ -694,7 +694,7 @@ inline void format_bam::write_alignment_record([[maybe_unused]] stream_type &  s
         if (ref_offset.has_value() && (ref_offset.value() + 1) < 0)
             throw format_error{detail::to_string("The ref_offset object must be >= -1 but is: ", ref_offset)};
 
-        seqan3::ostreambuf_iterator stream_it{stream};
+        std::cpp20::ostreambuf_iterator stream_it{stream};
```
Comment (smehringer) [inline include/seqan3/io/alignment_file/format_bam.hpp:697]: Why did you change this? We explicitly introduced our own ostreambuf_iterator because we had a significant performance improvement when exposing the buffer pointers and writing directly to the buffer.

Comment (smehringer) [reply include/seqan3/io/alignment_file/format_bam.hpp:697]: Edit: Sorry, I just saw that you are right `seqan3::ostreambuf_iterator` indeed is either ranges::ostreambuf_iterator or std::ostreambuf_iterator. It is the `seqan::fast_ostreambuf_iterator` I was talking about. I think it should be changed to  `seqan::fast_ostreambuf_iterator` then.

Comment (marehr) [reply include/seqan3/io/alignment_file/format_bam.hpp:697]: good that I renamed it to avoid confusion :)

Comment (marehr) [reply include/seqan3/io/alignment_file/format_bam.hpp:697]: This is a performance improvement and I would not want to do it in this PR. I think we should do measurements if we swap the implementations. I think this can be done by a student :)

Comment (smehringer) [reply include/seqan3/io/alignment_file/format_bam.hpp:697]: I think we should change it right away, why wait for an improvement. But rene is the reviwer so I'll leave this PR to him

File: include/seqan3/std/iterator (hunk 2 above)
Comment (rrahn) [reply include/seqan3/std/iterator:22]: > I normally write my reasons in the commit itself:

I am usually not going through the commits.

File: include/seqan3/io/alignment_file/format_bam.hpp (hunk 4 above)
Comment (rrahn) [reply include/seqan3/io/alignment_file/format_bam.hpp:697]: Yes and I/O is not part of this or the next quarter. So we can look at the performance when we work on that.

Comment (rrahn) [reply include/seqan3/io/alignment_file/format_bam.hpp:697]: I would really have the applications done with the other stuff that is on our plate.

File: include/seqan3/io/stream/iterator.hpp (hunk 3 above)
Comment (rrahn) [reply include/seqan3/io/stream/iterator.hpp:419]: Ok, thanks for clarifying.

File: include/seqan3/io/alignment_file/format_bam.hpp (hunk 4 above)
Comment (marehr) [reply include/seqan3/io/alignment_file/format_bam.hpp:697]: I'm not against it, because it is low-hanging fruit, but I think it is outside this PR.
